### PR TITLE
feat: Adding kubelogin capability to argocd-k8s-auth (#9460)

### LINKED ISSUE #9460: Kubelogin support in ArgoCD
# Summary

It would be nice if ArgoCD could add support for the [Kubelogin plugin](https://github.com/Azure/kubelogin) for AKS clusters.

# Motivation

For security reasons we have enabled Azure AD RBAC for all of our clusters. However, right now it is not possible to use Kubelogin for ArgoCD. Instead it uses the cluster admin credentials which bypasses the Azure AD RBAC check. 

# Proposal

Add an option to enable the Kubelogin plugin. Kubelogin is based on the client-go credential plugin: https://kubernetes.io/docs/reference/access-authn-authz/authentication/#client-go-credential-plugins

### PR DESCRIPTION
**Pull request by mmerrill3**
Fixes: #9460

Signed-off-by: mmerrill3 <jjpaacks@gmail.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by mmerrill3** [comment]
> Thanks @mmerrill3 !
> 
> May I ask where all the updates to dependencies in `go.mod` are coming from? If those bumps are not implicit of the changes you made, I'd really like to see them going in as a different PR.

Hi @jannfis, it was from running go mod tidy after adding the kubelogin package.  I need to rebase with conflicts for go.mod and go.sum, so let me just add the kubelogin package without a go mod tidy.

**Comment by mmerrill3** [comment]
Hi @jannfis, looks like there's a lot of overlap with argocd depdencies, and kubelogin at [here](https://github.com/Azure/kubelogin/blob/2657f536d32079df15750c6d8973d0f4ac7e30c5/go.mod)

I just added kubelogin, latest version, with go get, and this is what go updated.  I guess another option is to bundle the binary kubelogin.

**Comment by mmerrill3** [comment]
@MattiasAng thanks for the insight, makes total sense.  Default the value to the one in the link, but let others override it if they need to use something else. .

**Comment by mmerrill3** [comment]
rebased and pushed the change

**Comment by mmerrill3** [comment]
@crenshaw-dev Anything you'd like to see to associate this with a milestone?

**Comment by blairdrummond** [comment]
Any timeline on this? I'm very excited to see it merged!

**Review by OzgurBagci — APPROVED** [review]

**Comment by mmerrill3** [comment]
I'll rebase later today and resolve the conflicts.

**Comment by OzgurBagci** [comment]
Hi, is there any progress. We had to built a custom image in order to circumvent the lack of support. Azure is a widely use cloud provider and a big project like ArgoCD should fully support it.

**Comment by crenshaw-dev** [comment]
@OzgurBagci so many community contributions are pouring in, the core team is struggling to keep up with reviews.

The fact that you’re running this successfully is helpful. Always easier to hit merge when someone’s already tested.

**Comment by akselleirv** [comment]
Would be great to see this merged, I'm also in need of this feature!

**Comment by mmerrill3** [comment]
@crenshaw-dev , I'm working through the go.mod updates, will squash when its good.  I'm pinning to kubelogin at github.com/Azure/kubelogin/pkg/token@v0.0.20, and working through any other updates.

File: docs/operator-manual/declarative-setup.md
Comment (Ziphone) [inline docs/operator-manual/declarative-setup.md:656]: Think this is invalid, @mmerrill3. ´Env´ seems to be a ´map[string]string´, given the related struct.

Comment (Ziphone) [reply docs/operator-manual/declarative-setup.md:656]: Would be:
```suggestion
"env": {
   "AAD_ENVIRONMENT_NAME": "AzurePublicCloud",
   "AZURE_CLIENT_ID": "fill in client id",
   "AZURE_TENANT_ID": "fill in tenant id",
   "AZURE_FEDERATED_TOKEN_FILE": "/opt/path/to/federated_file.json",
   "AZURE_AUTHORITY_HOST": "https://login.microsoftonline.com/",
   "AAD_LOGIN_METHOD": "workloadidentity"
 },
```

Comment (mmerrill3) [reply docs/operator-manual/declarative-setup.md:656]: @Ziphone , thank you, you are correct, I will fix.

Comment (mmerrill3) [reply docs/operator-manual/declarative-setup.md]: just pushed up change

Comment (Ziphone) [reply docs/operator-manual/declarative-setup.md]: Thanks! Hopefully it will get released during 2.6 🤞

File: docs/operator-manual/declarative-setup.md
Comment (Ziphone) [inline docs/operator-manual/declarative-setup.md:655]: Think this is invalid object notation in JSON. Missing `{}` embracing the fields, and commas delimiting each field of the object.

Comment (mmerrill3) [reply docs/operator-manual/declarative-setup.md]: long week @Ziphone, I had YAML on the mind.  Updated PR

**Comment by crenshaw-dev** [comment]
I'm not able to personally prioritize this for 2.7. If you'd like to advocate to get another reviewer, it might help to drop into the contributors meeting: https://github.com/argoproj/argoproj#project-resources

Sorry I can't help this release cycle!

**Comment by SayakMukhopadhyay** [comment]
This PR has given me an interest in doing something similar for Digital Ocean Kubernetes Service. It uses the `doctl` binary and right now I kind of use a hacky way to get it to work in 2.5.x, which incidentally breaks in 2.6.x (related #12403 #12404). So, will a PR for DO be accepted? If so, I will start right away.

**Comment by mmerrill3** [comment]
rebased today to bring in the go.mod changes from the origin/master.

**Comment by mmerrill3** [comment]
will rebase again today

**Comment by mmerrill3** [comment]
rebased again against master.

**Comment by mmerrill3** [comment]
@crenshaw-dev any chance I get get this considered for a future milestone?  Do I need to go to put this PR on the agenda and call in to the office hours meeting?

**Comment by OzgurBagci** [comment]
I am no longer in the project that I originally used the feature for, but still checking out the progress time to time. It is so sad this is still waiting for a reviewer, the feature owner needed to rebase multiple times. Thank you for your effort @mmerrill3

**Comment by mmerrill3** [comment]
@OzgurBagci, looks like this has been given a milestone, so that's great.  I'll rebase again today.

**Comment by mmerrill3** [comment]
rebased today

**Comment by mmerrill3** [comment]
rebased today

**Comment by mmerrill3** [comment]
rebased.

**Comment by mmerrill3** [comment]
rebased

**Comment by crenshaw-dev** [comment]
@mmerrill3 you're an absolute saint for keeping this maintained so long. This _will_ be in 2.8-rc1. I've been slow to review, because I have no way to test. But untested code that's relatively isolated is fine.

@todaywasawesome volunteered to test in his env. I'll wait until Friday to give him a chance to take it for a spin. Either way, I'll merge before we cut 2.8-rc1 on Monday.

**Comment by mmerrill3** [comment]
@crenshaw-dev I'll rebase today, and glad to help.

**Review by todaywasawesome — APPROVED** [review]
I didn't get a full chance to test but I did review the spec and I think it looks good.

**Review by crenshaw-dev — APPROVED** [review]

**Comment by mmerrill3** [comment]
@crenshaw-dev, I had to rebase as the go.mod was updated in the master which prevented the auto merge b/c of the conflict.

**Comment by mmerrill3** [comment]
rebased today to take in latest go.mod and go.sum

**Comment by mmerrill3** [comment]
@crenshaw-dev, sorry, auto merge was disabled, we were in a loop, the master brach's go.sum and go.mod is moving rapidly.  The auto-merge was disabled b/c I had to resync.  Merging was disabled b/c of conflicts with go.sum and go.mod.

**Comment by crenshaw-dev** [comment]
Oof, missed the auto-merge failure. I'll get this into 2.8.

**Comment by crenshaw-dev** [comment]
Rebased. No one breathe until it merges. :-P

@mmerrill3, there will definitely be merge conflicts on the cherry-pick to 2.8, so one of us will have to manually open that PR. If you have time, that's great, 'cause then I can smash merge. If not, I'll open the PR and find a reviewer before 2.8 GA Monday. :-)

**Comment by mmerrill3** [comment]
@crenshaw-dev , I opened a cherry pick PR at #14866 .

I made a branch off of release-2.8 and cherry picked into that, resolved conflicts.
To resolve conflicts:

- reset go.mod and go.sum to what is in release-2.8 branch
- added kubelogin with go get github.com/Azure/kubelogin@v0.0.20
- ran go mod tidy
- git added the go.mod and go.sum and finished the cherry pick

**Pull request merged**
